### ISSUE TITLE
assembly data flow graph

### ISSUE THREAD

**Issue by Trass3r** [mattgodbolt/compiler-explorer#1158]
I just discovered the llvm-mca integration, that's awesome!
With the -timeline option you do get an idea of the instruction dependency chains:
https://godbolt.org/z/-cMrJI

But it would be nicer if there was a graph like the following, just on an MCInst level:
https://github.com/k3ut0i/llvm-dataflow-graphs
![flow graph](https://raw.githubusercontent.com/k3ut0i/llvm-dataflow-graphs/master/ctrl-data.png)

**Comment by mattgodbolt** [comment]
Agreed: I'd ideally love to parse llvm-mca output, or better still have it output parseable JSON of what happened. Then I can use any number of graph library layouts to try and show the information.

We already have CFG: https://godbolt.org/z/dFqf4H

**Comment by Trass3r** [comment]
Yeah it should really invoke something along the lines of `opt -dot-cfg` and then display the dot file.

**Comment by OfekShilon** [comment]
Please note that 
```
opt -passes="view-cfg" -cfg-func-name=<func> <ll filename> -disable-output
```

produces better graphs - colors encoding regions. Maybe useful.

**Comment by jeremy-rifkin** [comment]
@OfekShilon Is there an equivalent of this for data flow? It looks like opt is generating a control flow graph here

**Comment by OfekShilon** [comment]
> @OfekShilon Is there an equivalent of this for data flow? It looks like opt is generating a control flow graph here

No ,sorry. I was lead hear by a link from the suggestion regarding cfg + data-dep, and saw that the start of the discussion touched on cfg so thought I'd comment here.

Please note that IR cfg (as the opt passes dot-cfg or view-cfg emit) is not strictly equivalent to assembly cfg (which is already available). Personally I think both are useful.
